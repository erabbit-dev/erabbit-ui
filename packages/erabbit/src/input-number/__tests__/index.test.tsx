import { InputNumber } from '..'
import { mount } from '@vue/test-utils'
import { nextTick, ref } from 'vue'

describe('InputNumber.vue', () => {
  test('create', async () => {
    const num = ref(1)
    const wrapper = mount(() => <InputNumber v-model={num.value} />)
    expect(wrapper.find('input').exists()).toBe(true)
  })

  test('modelValue', () => {
    const num = ref(1)
    const wrapper = mount(() => <InputNumber v-model={num.value} />)
    expect(wrapper.find('input').element.valueAsNumber).toEqual(1)
  })

  test('set modelValue undefined to form validate', async () => {
    const num = ref(undefined)
    const wrapper = mount(() => (
      <>
        <InputNumber modelValue={num.value} />
        <p>{num.value}</p>
      </>
    ))
    await nextTick()
    expect(wrapper.find('p').element.innerText).toBeUndefined()
  })

  test('set modelValue undefined to display placeholder', async () => {
    const inputText = ref<number | undefined>(1)
    const wrapper = mount(() => <InputNumber modelValue={inputText.value} />)
    expect(wrapper.find('input').element.value).toEqual('1')
    inputText.value = undefined
    await nextTick()
    expect(wrapper.find('input').element.value).toEqual('')
  })

  test('Make sure the input action can trigger the modelValue update', async () => {
    const num = ref<number>(0)
    const handleUpdate = (data: number | undefined) => {
      num.value = data!
    }
    const wrapper = mount(() => (
      <InputNumber modelValue={num.value} onUpdate:modelValue={handleUpdate} />
    ))
    const el = wrapper.find('input').element
    const simulateEvent = (text: string, event: string) => {
      el.value = text
      el.dispatchEvent(new Event(event))
    }
    simulateEvent('3', 'input')
    await nextTick()
    expect(num.value).toEqual(3)
  })

  test('Make sure modelValue correct update when no initial value', async () => {
    const num = ref<number>()
    const wrapper = mount(() => <InputNumber v-model={num.value} />)
    const inputWrapper = wrapper.find('input')
    const nativeInput = inputWrapper.element
    nativeInput.value = 1
    await inputWrapper.trigger('input')
    nativeInput.value = ''
    await inputWrapper.trigger('input')
    expect(num.value).toEqual(NaN)
    wrapper.find('.er-input-number__decrease').trigger('click')
    await nextTick()
    expect(wrapper.find('input').element.valueAsNumber).toEqual(1)
  })

  test('min', async () => {
    const num = ref(2)
    const wrapper = mount(() => <InputNumber min={1} v-model={num.value} />)
    wrapper.find('.er-input-number__decrease').trigger('click')
    await nextTick()
    expect(wrapper.find('input').element.valueAsNumber).toEqual(1)
    wrapper.find('.er-input-number__decrease').trigger('click')
    await nextTick()
    expect(wrapper.find('input').element.valueAsNumber).toEqual(1)
    wrapper.find('.er-input-number__decrease').trigger('click')
  })

  test('max', async () => {
    const num = ref(1)
    const wrapper = mount(() => <InputNumber max={2} v-model={num.value} />)
    wrapper.find('.er-input-number__increase').trigger('click')
    await nextTick()
    expect(wrapper.find('input').element.valueAsNumber).toEqual(2)
    wrapper.find('.er-input-number__increase').trigger('click')
    await nextTick()
    expect(wrapper.find('input').element.valueAsNumber).toEqual(2)
  })

  test('step, increase and decrease', async () => {
    const num = ref(0)
    const wrapper = mount(() => <InputNumber v-model={num.value} step={2} />)
    wrapper.find('.er-input-number__decrease').trigger('click')
    await nextTick()
    expect(wrapper.find('input').element.valueAsNumber).toEqual(-2)
    wrapper.find('.er-input-number__increase').trigger('click')
    await nextTick()
    expect(wrapper.find('input').element.valueAsNumber).toEqual(0)
    wrapper.find('.er-input-number__increase').trigger('click')
    await nextTick()
    expect(wrapper.find('input').element.valueAsNumber).toEqual(2)
  })

  test('disabled', async () => {
    const num = ref(0)
    const wrapper = mount(() => (
      <InputNumber disabled={true} v-model={num.value} />
    ))
    wrapper.find('.er-input-number__decrease').trigger('click')
    await nextTick()
    expect(wrapper.find('input').element.valueAsNumber).toEqual(0)
    wrapper.find('.er-input-number__increase').trigger('click')
    await nextTick()
    expect(wrapper.find('input').element.valueAsNumber).toEqual(0)
  })

  test('change-event', async () => {
    const num = ref(0)
    const wrapper = mount(() => <InputNumber v-model={num.value} />)
    wrapper.find('.er-input-number__increase').trigger('click')
    await nextTick()
    expect(wrapper.getComponent(InputNumber).emitted('change')).toHaveLength(1)
    expect(wrapper.getComponent(InputNumber).emitted().change[0]).toEqual([
      1, 0,
    ])
    expect(
      wrapper.getComponent(InputNumber).emitted('update:modelValue'),
    ).toHaveLength(1)
    wrapper.find('.er-input-number__increase').trigger('click')
    await nextTick()
    expect(wrapper.getComponent(InputNumber).emitted('change')).toHaveLength(2)
    expect(wrapper.getComponent(InputNumber).emitted().change[1]).toEqual([
      2, 1,
    ])
  })

  test('blur-event', async () => {
    const num = ref(0)
    const wrapper = mount(() => <InputNumber v-model={num.value} />)
    await wrapper.find('input').trigger('blur')
    expect(wrapper.getComponent(InputNumber).emitted('blur')).toHaveLength(1)
  })

  test('focus-event', async () => {
    const num = ref(0)
    const wrapper = mount(() => <InputNumber v-model={num.value} />)
    await wrapper.find('input').trigger('focus')
    expect(wrapper.getComponent(InputNumber).emitted('focus')).toHaveLength(1)
  })
})
