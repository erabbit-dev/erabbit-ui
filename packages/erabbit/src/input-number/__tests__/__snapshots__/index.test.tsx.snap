// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render InputNumber with props 1`] = `
"<div class=\\"er-input-number disabled er-input-number--large\\"><button disabled=\\"\\" class=\\"er-input-number__decrease\\"><svg viewBox=\\"0 0 1024 1024\\" xmlns=\\"http://www.w3.org/2000/svg\\">
      <path fill=\\"currentColor\\" d=\\"M128 544h768a32 32 0 1 0 0-64H128a32 32 0 0 0 0 64z\\"></path>
    </svg></button><input class=\\"er-input-number__input\\" autocomplete=\\"off\\" type=\\"number\\" name=\\"input-number\\" min=\\"0\\" max=\\"10\\" step=\\"2\\" placeholder=\\"请输入数字\\" disabled=\\"\\"><button class=\\"er-input-number__increase\\" disabled=\\"\\"><svg viewBox=\\"0 0 1024 1024\\" xmlns=\\"http://www.w3.org/2000/svg\\">
      <path fill=\\"currentColor\\" d=\\"M480 480V128a32 32 0 0 1 64 0v352h352a32 32 0 1 1 0 64H544v352a32 32 0 1 1-64 0V544H128a32 32 0 0 1 0-64h352z\\"></path>
    </svg></button></div>"
`;

exports[`should render default InputNumber 1`] = `
"<div class=\\"er-input-number er-input-number--default\\"><button class=\\"er-input-number__decrease\\"><svg viewBox=\\"0 0 1024 1024\\" xmlns=\\"http://www.w3.org/2000/svg\\">
      <path fill=\\"currentColor\\" d=\\"M128 544h768a32 32 0 1 0 0-64H128a32 32 0 0 0 0 64z\\"></path>
    </svg></button><input class=\\"er-input-number__input\\" autocomplete=\\"off\\" type=\\"number\\" min=\\"-Infinity\\" max=\\"Infinity\\" step=\\"1\\"><button class=\\"er-input-number__increase\\"><svg viewBox=\\"0 0 1024 1024\\" xmlns=\\"http://www.w3.org/2000/svg\\">
      <path fill=\\"currentColor\\" d=\\"M480 480V128a32 32 0 0 1 64 0v352h352a32 32 0 1 1 0 64H544v352a32 32 0 1 1-64 0V544H128a32 32 0 0 1 0-64h352z\\"></path>
    </svg></button></div>"
`;
